File: tests/sample.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;

class User {
  final int id;
  final String name;
  final String email;

  const User({required this.id, required this.name, required this.email});

  factory User.fromJson(Map<String, dynamic> json) => User(
        id: json['id'] as int,
        name: json['name'] as String,
        email: json['email'] as String,
      );

  @override
  String toString() => 'User(id: $id, name: $name)';
}

class UserService {
  final String _baseUrl;
  final _cache = <int, User>{};

  UserService(this._baseUrl);

  Future<User?> getUser(int id) async {
    if (_cache.containsKey(id)) return _cache[id];

    final uri = Uri.parse('$_baseUrl/users/$id');
    final res = await http.get(uri);
    if (res.statusCode != 200) return null;

    final user = User.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
    _cache[id] = user;
    return user;
  }
}

void main() async {
  final svc = UserService('https://api.example.com');
  final user = await svc.getUser(1);
  print(user != null ? 'Hello, ${user.name}!' : 'User not found');
}

File: tests/sample.rs
use std::collections::HashMap;

#[derive(Debug, Clone)]
struct User {
    id: u32,
    name: String,
    email: String,
}

struct UserService {
    cache: HashMap<u32, User>,
}

impl UserService {
    fn new() -> Self {
        Self { cache: HashMap::new() }
    }

    fn get_user(&mut self, id: u32) -> Option<&User> {
        if !self.cache.contains_key(&id) {
            let user = self.fetch_from_db(id)?;
            self.cache.insert(id, user);
        }
        self.cache.get(&id)
    }

    fn fetch_from_db(&self, id: u32) -> Option<User> {
        match id {
            1 => Some(User { id: 1, name: "Alice".into(), email: "alice@example.com".into() }),
            2 => Some(User { id: 2, name: "Bob".into(),   email: "bob@example.com".into() }),
            _ => None,
        }
    }
}

fn main() {
    let mut svc = UserService::new();
    match svc.get_user(1) {
        Some(u) => println!("Hello, {}!", u.name),
        None    => println!("User not found"),
    }
}
